File: internal/store/metrics.go
package store

import (
	"fmt"
	"runtime"
	"strconv"
	"strings"
	"sync/atomic"
)

// responseCounts buckets OGC responses by status for Prometheus.
type responseCounts struct {
	ok200, ok204, ok304    atomic.Uint64
	err400, err404, err406 atomic.Uint64
	err429, err500         atomic.Uint64
}

// Metrics mirrors /metrics without consuming a pool connection, in
// Prometheus text exposition (valid numbers only).
func (s *Store) Metrics() string {
	var b strings.Builder
	b.WriteString("# HELP lakewing_http_responses_total OGC responses served by status.\n# TYPE lakewing_http_responses_total counter\n")
	for _, kv := range [][2]string{
		{"200", strconv.FormatUint(s.responses.ok200.Load(), 10)},
		{"204", strconv.FormatUint(s.responses.ok204.Load(), 10)},
		{"304", strconv.FormatUint(s.responses.ok304.Load(), 10)},
		{"400", strconv.FormatUint(s.responses.err400.Load(), 10)},
		{"404", strconv.FormatUint(s.responses.err404.Load(), 10)},
		{"406", strconv.FormatUint(s.responses.err406.Load(), 10)},
		{"429", strconv.FormatUint(s.responses.err429.Load(), 10)},
		{"500", strconv.FormatUint(s.responses.err500.Load(), 10)},
	} {
		fmt.Fprintf(&b, "lakewing_http_responses_total{status=%q} %s\n", kv[0], kv[1])
	}
	b.WriteString("# HELP lakewing_http_requests_total OGC request attempts.\n# TYPE lakewing_http_requests_total counter\n")
	fmt.Fprintf(&b, "lakewing_http_requests_total %d\n", s.httpRequests.Load())
	threads, memory := "", ""
	for _, kv := range s.tuning {
		switch kv[0] {
		case "threads":
			threads = kv[1]
		case "memory_limit":
			memory = kv[1]
		}
	}
	b.WriteString("# HELP lakewing_duckdb_threads Shared DuckDB threads.\n# TYPE lakewing_duckdb_threads gauge\n")
	fmt.Fprintf(&b, "lakewing_duckdb_threads %s\n", numericOrZero(threads))
	b.WriteString("# HELP lakewing_duckdb_memory_limit_bytes Shared DuckDB memory budget.\n# TYPE lakewing_duckdb_memory_limit_bytes gauge\n")
	fmt.Fprintf(&b, "lakewing_duckdb_memory_limit_bytes %d\n", parseBytes(memory))
	var mem runtime.MemStats
	runtime.ReadMemStats(&mem)
	b.WriteString("# HELP lakewing_go_goroutines Live goroutines.\n# TYPE lakewing_go_goroutines gauge\n")
	fmt.Fprintf(&b, "lakewing_go_goroutines %d\n", runtime.NumGoroutine())
	b.WriteString("# HELP lakewing_go_heap_bytes Go heap in use.\n# TYPE lakewing_go_heap_bytes gauge\n")
	fmt.Fprintf(&b, "lakewing_go_heap_bytes %d\n", mem.HeapInuse)
	return b.String()
}

// numericOrZero passes through plain integers, else 0 (Prometheus needs
// bare numbers).
func numericOrZero(s string) string {
	s = strings.TrimSpace(s)
	if _, err := strconv.Atoi(s); err != nil {
		return "0"
	}
	return s
}

// parseBytes parses DuckDB byte quantities ("1.0 GiB", "512MiB", "4096").
func parseBytes(s string) uint64 {
	s = strings.TrimSpace(s)
	mult := 1.0
	// Longest suffixes first so "GiB" wins over "B" (map order is random).
	for _, suf := range []struct {
		suffix string
		mult   float64
	}{
		{"TiB", 1 << 40}, {"GiB", 1 << 30}, {"MiB", 1 << 20}, {"KiB", 1 << 10},
		{"TB", 1e12}, {"GB", 1e9}, {"MB", 1e6}, {"KB", 1e3}, {"B", 1},
	} {
		if v, ok := strings.CutSuffix(s, suf.suffix); ok {
			s, mult = strings.TrimSpace(v), suf.mult
			break
		}
	}
	f, err := strconv.ParseFloat(s, 64)
	if err != nil {
		return 0
	}
	return uint64(f * mult)
}

func (s *Store) CountRequest() { s.httpRequests.Add(1) }

// CountResponse buckets one served OGC response by status.
func (s *Store) CountResponse(status int) {
	switch status {
	case 200:
		s.responses.ok200.Add(1)
	case 204:
		s.responses.ok204.Add(1)
	case 304:
		s.responses.ok304.Add(1)
	case 400:
		s.responses.err400.Add(1)
	case 404:
		s.responses.err404.Add(1)
	case 406:
		s.responses.err406.Add(1)
	case 429:
		s.responses.err429.Add(1)
	default:
		s.responses.err500.Add(1)
	}
}

File: crates/lakewing/src/metrics.rs
//! Prometheus text exposition over atomic counters (ported shape from the
//! Go store metrics; Go-runtime gauges replaced by in-flight and dataset
//! metadata).
use std::sync::atomic::{AtomicU64, Ordering};

pub struct Metrics {
    requests: AtomicU64,
    in_flight: AtomicU64,
    ok200: AtomicU64,
    ok204: AtomicU64,
    ok304: AtomicU64,
    err400: AtomicU64,
    err404: AtomicU64,
    err429: AtomicU64,
    err500: AtomicU64,
    pub dataset_version: AtomicU64,
}

impl Default for Metrics {
    fn default() -> Self {
        Self::new()
    }
}

impl Metrics {
    pub fn new() -> Self {
        Self {
            requests: AtomicU64::new(0),
            in_flight: AtomicU64::new(0),
            ok200: AtomicU64::new(0),
            ok204: AtomicU64::new(0),
            ok304: AtomicU64::new(0),
            err400: AtomicU64::new(0),
            err404: AtomicU64::new(0),
            err429: AtomicU64::new(0),
            err500: AtomicU64::new(0),
            dataset_version: AtomicU64::new(0),
        }
    }

    pub fn count_request(&self) {
        self.requests.fetch_add(1, Ordering::Relaxed);
    }

    pub fn enter(&self) {
        self.in_flight.fetch_add(1, Ordering::Relaxed);
    }

    pub fn leave(&self) {
        self.in_flight.fetch_sub(1, Ordering::Relaxed);
    }

    pub fn count(&self, status: u16) {
        match status {
            200 => self.ok200.fetch_add(1, Ordering::Relaxed),
            204 => self.ok204.fetch_add(1, Ordering::Relaxed),
            304 => self.ok304.fetch_add(1, Ordering::Relaxed),
            400 => self.err400.fetch_add(1, Ordering::Relaxed),
            404 => self.err404.fetch_add(1, Ordering::Relaxed),
            429 => self.err429.fetch_add(1, Ordering::Relaxed),
            _ => self.err500.fetch_add(1, Ordering::Relaxed),
        };
    }

    pub fn exposition(&self) -> String {
        let mut out = String::new();
        out.push_str("# HELP lakewing_http_responses_total OGC responses served by status.\n# TYPE lakewing_http_responses_total counter\n");
        for (status, value) in [
            ("200", self.ok200.load(Ordering::Relaxed)),
            ("204", self.ok204.load(Ordering::Relaxed)),
            ("304", self.ok304.load(Ordering::Relaxed)),
            ("400", self.err400.load(Ordering::Relaxed)),
            ("404", self.err404.load(Ordering::Relaxed)),
            ("429", self.err429.load(Ordering::Relaxed)),
            ("500", self.err500.load(Ordering::Relaxed)),
        ] {
            out.push_str(&format!(
                "lakewing_http_responses_total{{status=\"{status}\"}} {value}\n"
            ));
        }
        out.push_str("# HELP lakewing_http_requests_total OGC request attempts.\n# TYPE lakewing_http_requests_total counter\n");
        out.push_str(&format!(
            "lakewing_http_requests_total {}\n",
            self.requests.load(Ordering::Relaxed)
        ));
        out.push_str("# HELP lakewing_http_in_flight Requests currently executing.\n# TYPE lakewing_http_in_flight gauge\n");
        out.push_str(&format!(
            "lakewing_http_in_flight {}\n",
            self.in_flight.load(Ordering::Relaxed)
        ));
        out.push_str("# HELP lakewing_lance_dataset_version Pinned serving dataset version.\n# TYPE lakewing_lance_dataset_version gauge\n");
        out.push_str(&format!(
            "lakewing_lance_dataset_version {}\n",
            self.dataset_version.load(Ordering::Relaxed)
        ));
        out
    }
}
